### PR TITLE
Combine BindableNumber implementations into base class

### PR DESCRIPTION
**Pull request by peppy**
Generally makes working with `BindableNumber<>` a lot better.

### CODE REVIEW

File: osu.Framework/Bindables/BindableDouble.cs [hunk 1]
```diff
@@ -1,19 +1,14 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Globalization;
 
 namespace osu.Framework.Bindables
 {
+    [Obsolete("Use BindableNumber<int> instead")] // can be removed 20200609
```
Comment (smoogipoo) [inline osu.Framework/Bindables/BindableDouble.cs]:
```suggestion
    [Obsolete("Use BindableNumber<double> instead")] // can be removed 20200609
```

File: osu.Framework/Bindables/BindableFloat.cs [hunk 2]
```diff
@@ -1,19 +1,14 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Globalization;
 
 namespace osu.Framework.Bindables
 {
+    [Obsolete("Use BindableNumber<int> instead")] // can be removed 20200609
```
Comment (smoogipoo) [inline osu.Framework/Bindables/BindableFloat.cs]:
```suggestion
    [Obsolete("Use BindableNumber<float> instead")] // can be removed 20200609
```

File: osu.Framework.Tests/Visual/Bindables/TestSceneBindableNumbers.cs [hunk 3]
```diff
@@ -12,10 +12,10 @@ namespace osu.Framework.Tests.Visual.Bindables
 {
     public class TestSceneBindableNumbers : FrameworkTestScene
     {
-        private readonly BindableInt bindableInt = new BindableInt();
+        private readonly BindableNumber<int> bindableInt = new BindableNumber<int>();
         private readonly BindableLong bindableLong = new BindableLong();
```
Comment (smoogipoo) [inline osu.Framework.Tests/Visual/Bindables/TestSceneBindableNumbers.cs]: `BindableLong` seems to be used, but not have appropriate `Default*` implementations.

**Comment by peppy** [comment]
As discussed IRL, won't be obsoleting usages due to them being shorter and easier on the eye.

**Comment by smoogipoo** [comment]
Should remove the overridden ToString() implementations such as https://github.com/ppy/osu-framework/pull/3089/files#diff-1119be848d4dba156f3cd1030d3e87f7R15

**Comment by peppy** [comment]
Intentionally left those for parity with master.

File: osu.Framework/Bindables/BindableNumber.cs [hunk 4]
```diff
@@ -112,17 +112,65 @@ public T MaxValue
...
+            get
+            {
+                if (typeof(T) == typeof(double))
+                    return (T)(object)double.Epsilon;
+                if (typeof(T) == typeof(float))
+                    return (T)(object)float.Epsilon;
+                if (typeof(T) == typeof(int))
+                    return (T)(object)1;
+                if (typeof(T) == typeof(long))
+                    return (T)(object)1;
```
Comment (huoyaoyuan) [inline osu.Framework/Bindables/BindableNumber.cs]: `(T)(object)1L`
`1` is int32, and can't be unboxed into long.

File: osu.Framework/Bindables/BindableNumber.cs [hunk 5]
```diff
@@ -293,6 +341,28 @@ public void SetProportional(float amt, float snap = 0)
 
         public new BindableNumber<T> GetUnboundCopy() => (BindableNumber<T>)base.GetUnboundCopy();
 
+        public override string ToString() => Value.ToString(NumberFormatInfo.InvariantInfo);
+
+        public override bool IsDefault
+        {
+            get
+            {
+                switch (Value)
```
Comment (huoyaoyuan) [inline osu.Framework/Bindables/BindableNumber.cs]: Why not checking `typeof(T)` here? Using switch may cause boxing.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4aNANoApcRgDiMETCjiwACgwBPDjAgAzXQBMIQ4ABsYASgcBdGsQDMDMgwDCAHgAVAD4GAG8aBkivTwCGADVsWyEYMIBzGAxpXAzpAF8IqILIjwZxMQYAWV0HIrDaqIZcAHcVMBZdBKTHeqjw6gaBqLA8FKsbewYLEB7BhuIAdgZdMowHC1l+2Ybh7IZTWwhsDD3pza3BhaWVh1MN86iLGFNsIVssGfPL8lJ3O9n8s4MAG5IA=

Comment (smoogipoo) [reply osu.Framework/Bindables/BindableNumber.cs]: With that being said there are two more boxings in each pathway. Can you think of a way to remove those too?

Comment (huoyaoyuan) [reply osu.Framework/Bindables/BindableNumber.cs]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4aNANoApcRgDiMETCjiwACgwBPDjAgAzXQBMIQ4ABsYASgcBdGsQDMDMgwDCAHgAVAD4GAG8aBkivTwCGADVsWyEYMIBzGAxpXAzpAF8IqILIjwZxMQYAWV0HIrDaqNLTBn0jE3MAhwYAXi6GQ2MzS2s7RxrqBomo4gB2ZrKMByGbe0WIYAArGDAFhKSZeqiYW2zG5v623Q7u3vPB01sIbAWxycmZubFF+8eF3TXN7YOXbJWTjV5HbIHN6zcikdyghr5ai5IA=
JIT knows the pattern `(ValueType)(object)T`, and will optimize away it.

Comment (smoogipoo) [reply osu.Framework/Bindables/BindableNumber.cs]: Perfect. TIL.

**Pull request merged**